### PR TITLE
Task filter

### LINKED ISSUE #509: Smart Search filter for tasks
## Description
A user should be able to configure a Smart Search to include the new task-based filter, to search for people based on their task information.

## Relevant Job Stories
N/A

## Prerequisites
Requires the task filter to be implemented in the platform first, which is currently in progress.

## Requirements
- [ ] The new filter should be selectable in the Smart Search dialog
- [ ] The new filter should be configurable (all variations) in the Smart Search dialog

## Possible implementations
Look at how other filters have been implemented, and reuse widgets when applicable.

## Design specifications
N/A

## Open questions
None at this time.

### PR DESCRIPTION
**Pull request by niklasva82**
## Description
This PR adds a task filter to the smart search.

## Screenshots

Screenshot from 2022-01-21 08-09-37
Screenshot from 2022-01-21 08-18-24
Screenshot from 2022-01-21 08-18-41

## Changes
[Add a list of features added/changed, bugs fixed etc]

* Adds a component for editing task type smart search filters.
* Adds a component for viewing task type smart search filters.
* Adds a component for editing `matching` type filters with `max` and `min` matches.

## Notes to reviewer
Add a new filter and try out the various combinations. Be sure to use the platform PR for fixed tasks filter.

## Related issues
Resolves #509

### CODE REVIEW

**Comment by richardolsson** [comment]
Playwright tests are failing. I don't know what's causing it, but might be fixed by the recently merged changes to `main`. Can you merge in the most recent `main` again and see if that fixes it?

File: src/locale/misc/smartSearch/en.yml
Comment (richardolsson) [inline src/locale/misc/smartSearch/en.yml:146]: This should be "been assigned".
image

File: src/types/smartSearch.ts
Comment (richardolsson) [inline src/types/smartSearch.ts:194]: Nice! This actually (much better) expresses in code what I was trying to say the other day, which is that even without the `boolean` this is still a union type.

File: src/types/smartSearch.ts
Comment (richardolsson) [inline src/types/smartSearch.ts:203]: There's a type for this union I believe, or rather an enum. Check out `TASK_TYPE` in `src/types/tasks.ts`.

File: src/types/smartSearch.ts
Comment (richardolsson) [inline src/types/smartSearch.ts:15]: Here you're calling it `task`, but in the platform it seems to be called `tasks`, so I can't actually get the GUI to work with the platform API. Is it working for you?

**Review by richardolsson — CHANGES_REQUESTED** [review]
I think this is mostly looking great, but it's not quite working and there are a couple of things that need to changed in the code before a final review. Comments below 👇

**Comment by richardolsson** [comment]
Having played some more with the filter, it seems that it will often generate configs with `matching: {}`, `assigned: {}` and similar constructs that should no longer be possible, right? Are you able to reproduce this or do you need me to specify steps to reproduce? Just picking "are assigned to" and then submitting seems to be enough to trigger the incorrect `{}` value for `assigned`.

IMO, if `matching` is empty, it should just be removed. If one does not care about the number of tasks, then `matching` is irrelevant and should not be included in the config.

**Comment by niklasva82** [comment]
`matching: {}` doesn't matter, `{}` is possible in the API. But `assigned: {}`, etc. should never happen now.

File: src/locale/misc/smartSearch/en.yml
Comment (niklasva82) [reply src/locale/misc/smartSearch/en.yml]: fixed

File: src/types/smartSearch.ts
Comment (niklasva82) [reply src/types/smartSearch.ts]: Yeah I figured this would help to find any locations where this was not the case, in order to eliminate any `{}`. Obviously didn't work out as well as I had hoped though.

File: src/types/smartSearch.ts
Comment (niklasva82) [reply src/types/smartSearch.ts]: fixed

File: src/types/smartSearch.ts
Comment (niklasva82) [reply src/types/smartSearch.ts:16]: Hm, I believe I fixed that on the platform side?

Comment (niklasva82) [reply src/types/smartSearch.ts:16]: Ah yeah, the filter validator called it task but the service used tasks. It's been fixed now in zetkin-platform#433

File: src/locale/misc/smartSearch/en.yml
Comment (richardolsson) [inline src/locale/misc/smartSearch/en.yml:146]: I think this should just be "been assigned", not "been assigned to", but let's consult a native English speaker like @djbusstop or @lowlandjuju. 😊 

Should it be "people who have been assigned any task…" or "people who have been assigned to any task…"?

image

**Review by richardolsson — COMMENTED** [review]
I think this is working great now! While waiting for the backend to be approved and merged, I have a question for our native English speaker friends. 👇 😊

File: src/locale/misc/smartSearch/en.yml
Comment (djbusstop) [reply src/locale/misc/smartSearch/en.yml:146]: I don't think it really matters either way, but I guess the best one is "have been assigned" without the word "to"

Comment (niklasva82) [reply src/locale/misc/smartSearch/en.yml]: Alright I updated to "been assigned". The shorter the better!

**Review by richardolsson — APPROVED** [review]
Green light to merge! 🟢  😊

**Pull request merged**
